### PR TITLE
async grpc: make grpc client eviction time configurable

### PR DESCRIPTION
**Pull request by botengyao**
Commit Message: async grpc: make grpc client eviction time configurable 
Additional Description:
Make the cached grpc client eviction time configurable, and before it is fixed 50s.

Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/common/grpc/async_client_manager_impl_test.cc
Comment (adisuissa) [inline test/common/grpc/async_client_manager_impl_test.cc:248]: Please add a one-line comment describing the test.

File: test/common/grpc/async_client_manager_impl_test.cc
Comment (adisuissa) [inline test/common/grpc/async_client_manager_impl_test.cc:256]: The refactor here causes some confusion, because a default async-client-mgr is always created, and then it is overridden here.
A better design will probably be having an `initialize(Opt<Config>)` method that needs to be called to initialize the async-client-mgr correctly.

File: test/common/grpc/async_client_manager_benchmark.cc
Comment (adisuissa) [inline test/common/grpc/async_client_manager_benchmark.cc:42]: Is this being used anywhere?

**Review by adisuissa — COMMENTED** [review]
Thanks! Left a few comments after a first pass.

File: test/common/grpc/async_client_manager_impl_test.cc
Comment (botengyao) [reply test/common/grpc/async_client_manager_impl_test.cc:255]: done

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (botengyao) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Right, changed the name.

File: source/common/grpc/async_client_manager_impl.cc
Comment (botengyao) [reply source/common/grpc/async_client_manager_impl.cc]: done

File: source/common/grpc/async_client_manager_impl.cc
Comment (botengyao) [reply source/common/grpc/async_client_manager_impl.cc]: Done, to DefaultEntryIdleDuration

File: source/common/grpc/async_client_manager_impl.cc
Comment (botengyao) [reply source/common/grpc/async_client_manager_impl.cc]: ops, it is a test log, removed.

File: source/common/grpc/async_client_manager_impl.cc
Comment (botengyao) [reply source/common/grpc/async_client_manager_impl.cc]: done

File: source/common/protobuf/utility.h
Comment (botengyao) [reply source/common/protobuf/utility.h]: good point, and changed to milliseconds.

File: test/common/grpc/async_client_manager_benchmark.cc
Comment (botengyao) [reply test/common/grpc/async_client_manager_benchmark.cc]: changed the way.

File: test/common/grpc/async_client_manager_impl_test.cc
Comment (botengyao) [reply test/common/grpc/async_client_manager_impl_test.cc]: Yes, my initial thought is to avoid many initialize() calls for other tests. Modified.

**Review by botengyao — COMMENTED** [review]
Thanks for the review!

File: changelogs/current.yaml
Comment (adisuissa) [inline changelogs/current.yaml:152]: nit: the field is `max_cached_entry_idle_duration`, so it should be the one described here.

File: source/common/grpc/async_client_manager_impl.cc
Comment (adisuissa) [inline source/common/grpc/async_client_manager_impl.cc:18]: nit: please move into the unnamed namespace if possible.

File: test/common/grpc/async_client_manager_impl_test.cc
Comment (adisuissa) [inline test/common/grpc/async_client_manager_impl_test.cc:35]: This repeats the same pattern as the other test (initializing and then overriding).
RE the need to call initialization on each method:
1. Due to the change of the manager from a non-pointer to a pointer, each test that uses that also needs to be modified, so adding the call to initialize() won't change much.
2. An alternative is to override the Test class (or a shared ancestor), pass the config as a parameter to the c'tor, and then add the new tests to the new Test class. (this should avoid calling initialize on each test).

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (adisuissa) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:141]: Consider adding a minimal value that is much larger than 0.
I think that at least 5 seconds makes sense here, but willing to hear others thoughts.
FWIW, it is always possible decreasing this value in the future without breaking anyone, but increasing it may break some user.

**Review by adisuissa — COMMENTED** [review]
Overall LGTM, mostly minor comments.
/assign-from @envoyproxy/senior-maintainers

**Review by lizan — COMMENTED** [review]
@botengyao can you elaborate what's the issue you're trying to resolve by make this configurable? Just for the background.

The change itself LGTM modulo those minor comments.

File: changelogs/current.yaml
Comment (botengyao) [reply changelogs/current.yaml]: done.

File: source/common/grpc/async_client_manager_impl.cc
Comment (botengyao) [reply source/common/grpc/async_client_manager_impl.cc]: done

File: test/common/grpc/async_client_manager_impl_test.cc
Comment (botengyao) [reply test/common/grpc/async_client_manager_impl_test.cc]: Changed to the initialize() way to keep the test readable.

File: api/envoy/config/bootstrap/v3/bootstrap.proto
Comment (botengyao) [reply api/envoy/config/bootstrap/v3/bootstrap.proto]: Good point, +1, and decreasing this value will not break in future.

**Comment by botengyao** [comment]
> @botengyao can you elaborate what's the issue you're trying to resolve by make this configurable? Just for the background.
> 
> The change itself LGTM modulo those minor comments.

Thank @lizan for the review! The background is we want to have longer cached duration to save the latency like pass some info through rpc, for the very low qps scenario.

**Review by lizan — APPROVED** [review]

**Pull request merged**
